Adds material on *, + and ? shortcuts

apblack opened the pull request:
These commits document the *, + and ? shortcuts, and updates the idioms chapter to describe how to build an AST containing a collection when your grammar describes a list.

apblack committed:
Changes summary of what follows.

apblack committed:
Amplifies text on matching methods.

apblack committed:
Adds +, * and ? shortcuts to Parser section.

apblack committed:
Describes grammar idioms as they actually work.

apblack committed:
Adds .xcworkspace to .gitignore

apblack committed:
Rewords sentences to avoid bad line breaks.

apblack committed:
Notes that the Moose version does not support the shortcuts.

Ducasse commented:
Andrew

I would like to keep the examples without the shortcuts because I extracted them from reading the grammars and it was painful.
Do you have them somewhere? Because I would like to avoid to lose this information. 

Stef

apblack commented:
> On 14 Sep 2017, at 20:58 , StéphaneDucasse <notifications@github.com> wrote:
> 
> Andrew
> 
> I would like to keep the examples without the shortcuts because I extracted them from reading the grammars and it was painful.
> Do you have them somewhere? Because I would like to avoid to lose this information.

Some are still there, I think.  Others I took out, because they didn’t work.  At least, I could not get them to work.  I send most of a day figuring out how to make work what I needed, because the stuff that I copied from the booklet would not work for me.

So,  if I have taken out stuff that is important, we should put it back, but also make sure that it is correct!

	Andrew

apblack committed:
Puts Stéph's examples of list-formation back into idioms chapter.

apblack committed:
Adds note about whitespace in REs.

apblack commented:
I've added back a (slightly simplified version of) your example of a comma-separated list using recursion.  In the process I also found out why it didn't work for me, and documented that.

This PR now also includes a note about apaces being significant inside [ and ] in REs.

apblack committed:
Emphasizes definition before use for scanner tokens.

apblack committed:
Emphasizes that token action methods must return a token.
Adds examples.  Changes note on Unicode to refer to the
directive, and not a checkbox that is no longer there.
